### PR TITLE
[IMP] core: validate constraints with sudo()

### PR DESCRIPTION
**Pull request by kmagusiak**
Check constraints in sudo because the check may be done in a different environment from which the values were changed. This is similar to computed-stored fields which are run in compute_sudo=True.

task-3484019
https://github.com/odoo/enterprise/pull/69943

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/survey/models/survey_survey.py [hunk 1]
```diff
@@ -427,11 +427,10 @@ def _check_survey_responsible_access(self):
                 - and there is a survey responsible,
                 - and this responsible is not survey manager (just survey officer),
             check the responsible is part of the list."""
-        surveys_to_check = self.filtered(lambda s: bool(s.user_id - s.restrict_user_ids))
-        if surveys_to_check:
-            valid_surveys = surveys_to_check._filtered_access("write")
-            failing_surveys_sudo = (self - valid_surveys).sudo()
-            if failing_surveys_sudo:
+        for user_id, surveys in self.filtered(lambda s: bool(s.user_id - s.restrict_user_ids)).grouped('user_id').items():
+            accessible = surveys.with_env(surveys.env(user=user_id, su=False))._filtered_access("write")
```
Comment (ryv-odoo) [inline addons/survey/models/survey_survey.py:431]: Use with_user instead ?

File: addons/survey/models/survey_survey.py [hunk 2]
```diff
@@ -427,11 +427,10 @@ def _check_survey_responsible_access(self):
...
+            if len(accessible) < len(surveys):
+                failing_surveys_sudo = (self - accessible).sudo()
```
Comment (ryv-odoo) [inline addons/survey/models/survey_survey.py:433]: self is already in sudo then the extra sudo isn't need IMO

File: odoo/models.py [hunk 3]
```diff
@@ -1590,12 +1590,18 @@ def _validate_fields(self, field_names, excluded_names=()):
         """ Invoke the constraint methods for which at least one field name is
         in ``field_names`` and none is in ``excluded_names``.
         """
+        methods = self._constraint_methods
+        if not methods:
+            return
+        # run constrains just as computed-stored fields
```
Comment (ryv-odoo) [inline odoo/models.py:1596]: sudoed compute-stored field ?

File: addons/survey/models/survey_survey.py (hunk 2 above)
Comment (kmagusiak) [reply addons/survey/models/survey_survey.py:433]: Indeed, now it is in sudo, but for readability (and if you were to call the function directly), I would leave it here.

**Review by ryv-odoo — APPROVED** [review]
@robodoo r+
Thx

**Pull request closed without merging**